### PR TITLE
feat: Bill of entry for Import of Goods

### LINKED ISSUE #463: Not able to book Purchase Invoice for unregistered and local vendor with IGST
### Information about bug

When import purchase is done companies get the IGST input which generally comes in CHA invoice or custom department and these suppliers are unregistered vendors or local vendors but the system doesn't allow booking IGST invoices.

One more use case for booking IGST is for getting input against Script DGFT

### Version

ERPNext Version - 14

### Relevant log output / Stack trace / Full Error Message.

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by vorasmit**
## Summary of changes made

- Bill of Entry (as new doctype) with Accounting Dimensions
- Create accounting dimensions fields if they already exist
- Create a Bill of entry from the Purchase Invoice
- Create Payment Entry (as journal entry) from Bill of entry
- Independent taxes table (for BOE) with its calculations (with item-wise rates).
- Default customs account field in Company
- GSTR 3B report with Import purchases

## Demo

https://user-images.githubusercontent.com/10496564/216298062-25d71cc4-cf22-4643-91ed-0efe9634a611.mp4

Closes: #463 
Review sponsored by [Windmill Fans](https://windmillfans.com/)

### CODE REVIEW

File: india_compliance/gst_india/setup/__init__.py [hunk 1]
```diff
@@ -68,6 +85,16 @@ def delete_custom_fields():
     )
 
 
+def delete_accounting_dimension_fields():
```
Comment (sagarvora) [inline india_compliance/gst_india/setup/__init__.py]: put in uninstall.py

File: india_compliance/gst_india/constants/custom_fields.py [hunk 2]
```diff
@@ -43,7 +43,16 @@
 company_fields[0]["insert_after"] = "parent_company"
 
 CUSTOM_FIELDS = {
-    "Company": company_fields,
+    "Company": [
+        *company_fields,
+        {
```
Comment (sagarvora) [inline india_compliance/gst_india/constants/custom_fields.py:53]: append instead?

File: india_compliance/gst_india/overrides/purchase_invoice.py [hunk 3]
```diff
@@ -43,3 +43,12 @@ def validate_supplier_gstin(doc):
             _("Supplier GSTIN and Company GSTIN cannot be the same"),
             title=_("Invalid Supplier GSTIN"),
         )
+
+
+def onload(doc, method):
+    existing_bill_of_entry = frappe.db.get_value(
```
Comment (sagarvora) [inline india_compliance/gst_india/overrides/purchase_invoice.py:52]: is this really needed? what happens with links in ERPNext doctypes? if needed, add conditions for docstatus and gst category.

File: india_compliance/gst_india/constants/custom_fields.py [hunk 4]
```diff
@@ -43,7 +43,16 @@
...
+            "fieldname": "default_customs_duty_account",
+            "label": "Default Customs Duty Account",
+            "fieldtype": "Link",
+            "options": "Account",
+            "insert_after": "unrealized_profit_loss_account",
+        },
+    ],
```
Comment (sagarvora) [inline india_compliance/gst_india/constants/custom_fields.py:72]: add field for default custom payable account. consider creating these accounts by default for new sites.

File: india_compliance/gst_india/doctype/bill_of_entry_taxes/bill_of_entry_taxes.json [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+  },
+  {
+   "columns": 2,
+   "fieldname": "total",
+   "fieldtype": "Currency",
+   "in_list_view": 1,
+   "label": "Total",
+   "oldfieldname": "total",
+   "oldfieldtype": "Currency",
+   "options": "Company:company:default_currency",
```
Comment (sagarvora) [inline india_compliance/gst_india/doctype/bill_of_entry_taxes/bill_of_entry_taxes.json]: hardcode INR?

File: india_compliance/gst_india/doctype/bill_of_entry/bill_of_entry.json [hunk 6]
```diff
@@ -0,0 +1,293 @@
...
+  "column_break_kgnb",
+  "total_amount_payable"
+ ],
+ "fields": [
+  {
+   "fieldname": "naming_series",
+   "fieldtype": "Select",
+   "label": "Naming Series",
+   "no_copy": 1,
+   "options": "GST-BOE-.YYYY.-",
```
Comment (sagarvora) [inline india_compliance/gst_india/doctype/bill_of_entry/bill_of_entry.json]:
```suggestion
   "options": "BOE-.YYYY.-",
```

File: india_compliance/gst_india/doctype/bill_of_entry_taxes/bill_of_entry_taxes.json [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+   "read_only": 1
+  },
+  {
+   "fieldname": "item_wise_tax_rates",
+   "fieldtype": "Code",
+   "label": "Item Wise Tax Rates"
```
Comment (sagarvora) [inline india_compliance/gst_india/doctype/bill_of_entry_taxes/bill_of_entry_taxes.json:91]: hidden!

File: india_compliance/gst_india/doctype/bill_of_entry/test_bill_of_entry.py [hunk 8]
```diff
@@ -0,0 +1,9 @@
+# Copyright (c) 2023, Resilient Tech and Contributors
+# See license.txt
+
+# import frappe
+from frappe.tests.utils import FrappeTestCase
+
+
+class TestBillofEntry(FrappeTestCase):
+    pass
```
Comment (sagarvora) [inline india_compliance/gst_india/doctype/bill_of_entry/test_bill_of_entry.py]: add test to create BoE.

**Review by sagarvora — CHANGES_REQUESTED** [review]

File: india_compliance/gst_india/overrides/purchase_invoice.py (hunk 3 above)
Comment (vorasmit) [reply india_compliance/gst_india/overrides/purchase_invoice.py:52]: Conditions added.

File: india_compliance/gst_india/doctype/bill_of_entry/bill_of_entry.py [hunk 9]
```diff
@@ -0,0 +1,504 @@
...
+        )
+
+        if existing_journal_entry:
+            self.set_onload("existing_journal_entry", existing_journal_entry)
+
+    def before_validate(self):
+        self.set_item_wise_tax_rates()
+        self.calculate_totals()
+
+    def before_submit(self):
```
Comment (sagarvora) [inline india_compliance/gst_india/doctype/bill_of_entry/bill_of_entry.py:40]: This should be "validate".

File: india_compliance/gst_india/doctype/bill_of_entry/bill_of_entry.py [hunk 10]
```diff
@@ -0,0 +1,504 @@
...
+            {
+                "reference_type": "Bill of Entry",
+                "reference_name": self.name,
+                "docstatus": 1,
+            },
+            "name",
+        )
+
+        if existing_journal_entry:
+            self.set_onload("existing_journal_entry", existing_journal_entry)
```
Comment (sagarvora) [inline india_compliance/gst_india/doctype/bill_of_entry/bill_of_entry.py:34]: [optional]

we aren't really using the values in `existing_journal_entry` or `existing_bill_of_entry`. In such cases, using `frappe.db.exists` coupled with a variable name like `journal_entry_exists` better conveys what we are trying to achieve.

```py
        self.set_onload("journal_entry_exists", frappe.db.exists...)
```

File: india_compliance/gst_india/doctype/bill_of_entry/bill_of_entry.py [hunk 11]
```diff
@@ -0,0 +1,504 @@
...
+    def on_submit(self):
+        self.create_gl_entries()
+
+    def on_cancel(self):
+        self.ignore_linked_doctypes = ("GL Entry",)
+        self.cancel_gl_entries()
+
+    def on_trash(self):
+        controller = AccountsController
+        controller.on_trash(self)
```
Comment (sagarvora) [inline india_compliance/gst_india/doctype/bill_of_entry/bill_of_entry.py:53]: A better approach:

Adding method as "own":
Screenshot-2023-03-23-113131

Usage (normal):
Screenshot-2023-03-23-113148

**Review by sagarvora — COMMENTED** [review]
We can disable the "new"  button in list view by checking "user cannot create" in doctype.

**Pull request merged**